envoy/cilium_host_map.cc

```diff
@@ -11,12 +11,124 @@
 namespace Envoy {
 namespace Cilium {
 
-PolicyHostMap::PolicyHostMap(std::unique_ptr<Envoy::Config::Subscription<cilium::NetworkPolicyHosts>>&& subscription,
-				   ThreadLocal::SlotAllocator& tls)
-  : tls_(tls.allocateSlot()), subscription_(std::move(subscription)) {
+template <typename T>
+unsigned int checkPrefix(T addr, unsigned int plen, absl::string_view host) {
+  const unsigned int PLEN_MAX = sizeof(T)*8;
+  if (plen > PLEN_MAX) {
+    throw EnvoyException(fmt::format("NetworkPolicyHosts: Invalid prefix length in \'{}\'", host));
+  }
+  if (plen == 0) {
+    return PLEN_MAX;
+  }
+  // Check for 1-bits after the prefix
+  T mask = (T(1) << (PLEN_MAX - plen)) - 1;
+  if (addr & ntoh(mask)) {
+    throw EnvoyException(fmt::format("NetworkPolicyHosts: Non-prefix bits set in \'{}\'", host));
+  }
+  return plen;
+}
+
+struct ThreadLocalHostMapInitializer : public PolicyHostMap::ThreadLocalHostMap,
+				       public Logger::Loggable<Logger::Id::config> {
+protected:
+  friend class PolicyHostMap; // PolicyHostMap can insert();
+
+  // find the map of the given prefix length, insert in the decreasing order if it does
+  // not exist
+  template <typename M>
+  M& getMap(std::vector<std::pair<unsigned int, M>>& maps, unsigned int plen) {
+    auto it = maps.begin();
+    for (; it != maps.end(); it++) {
+      if (it->first > plen) {
+	ENVOY_LOG(debug, "Skipping map for prefix length {} while looking for {}", it->first, plen);
```

Comment:
The debug logs in this method are going to be VERY verbose. How about using `trace` instead?

Reply:
Good point. I added them for tracing purposes anyway!